File: qc-core/src/main/java/com/eager/questioncloud/core/domain/payment/infrastructure/QuestionPaymentHistoryRepositoryImpl.java
package com.eager.questioncloud.core.domain.payment.infrastructure;

import com.eager.questioncloud.core.common.PagingInformation;
import com.eager.questioncloud.core.domain.payment.model.QuestionPaymentHistory;
import java.util.List;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Repository;

@Repository
@RequiredArgsConstructor
public class QuestionPaymentHistoryRepositoryImpl implements QuestionPaymentHistoryRepository {
    private final QuestionPaymentHistoryMongoRepository questionPaymentHistoryMongoRepository;

    @Override
    public List<QuestionPaymentHistory> getQuestionPaymentHistory(Long userId, PagingInformation pagingInformation) {
        return questionPaymentHistoryMongoRepository.findByUserId(
                userId,
                PageRequest.of(pagingInformation.getPage(), pagingInformation.getSize(), Sort.by(Sort.Order.desc("paymentId")))
            )
            .stream()
            .map(QuestionPaymentHistoryDocument::toModel)
            .collect(Collectors.toList());
    }

    @Override
    public int count(Long userId) {
        return questionPaymentHistoryMongoRepository.countByUserId(userId);
    }

    @Override
    public QuestionPaymentHistory save(QuestionPaymentHistory questionPaymentHistory) {
        return questionPaymentHistoryMongoRepository.save(QuestionPaymentHistoryDocument.from(questionPaymentHistory)).toModel();
    }
}

File: qc-infra/qc-mongo/src/main/kotlin/com/eager/questioncloud/payment/repository/QuestionPaymentHistoryRepositoryImpl.kt
package com.eager.questioncloud.payment.repository

import com.eager.questioncloud.common.pagination.PagingInformation
import com.eager.questioncloud.payment.document.QuestionPaymentHistoryDocument
import com.eager.questioncloud.payment.domain.QuestionPaymentHistory
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Repository
import java.util.stream.Collectors

@Repository
class QuestionPaymentHistoryRepositoryImpl(
    private val questionPaymentHistoryMongoRepository: QuestionPaymentHistoryMongoRepository
) : QuestionPaymentHistoryRepository {
    override fun getQuestionPaymentHistory(
        userId: Long,
        pagingInformation: PagingInformation
    ): List<QuestionPaymentHistory> {
        return questionPaymentHistoryMongoRepository.findByUserId(
            userId,
            PageRequest.of(pagingInformation.page, pagingInformation.size, Sort.by(Sort.Order.desc("paymentId")))
        )
            .stream()
            .map { document -> document.toModel() }
            .collect(Collectors.toList())
    }
    
    override fun count(userId: Long): Int {
        return questionPaymentHistoryMongoRepository.countByUserId(userId)
    }
    
    override fun save(questionPaymentHistory: QuestionPaymentHistory): QuestionPaymentHistory {
        return questionPaymentHistoryMongoRepository.save(
            QuestionPaymentHistoryDocument.from(
                questionPaymentHistory
            )
        )
            .toModel()
    }
}
